### PR TITLE
refactor: Update nearest-point algorithm to more explicitly use vector math

### PR DESCRIPTION
**Pull request by joshlarson**

### CODE REVIEW

File: lib/util/location.ex
Comment (joshlarson) [inline lib/util/location.ex:204]: `point_vector` seems like a silly name, but I'm not sure of a better one. It's the vector pointing from the `start` point of the segment to the non-segment point.

File: lib/util/vector2d.ex
Comment (joshlarson) [inline lib/util/vector2d.ex:4]: This needs docs and tests

File: lib/util/location.ex
Comment (firestack) [inline lib/util/location.ex:130]: just a note, I'm having difficulty understanding this in context. I understand what it's doing, but when I read `segment_vector = displacement(start_loc, end_loc)` it just doesn't make sense to me from reading it, and I've gotta do extra "mental 'math'" to figure out what I'm supposed to understand. Do you think there could be a different word to accurately describe this, or a way I should be thinking about this word?

Comment (joshlarson) [reply lib/util/location.ex:130]: Let's ideate tomorrow. I've been steeped in this for a while now, so it makes probably _too much sense_ to me 😅

File: lib/util/vector2d.ex
Comment (joshlarson) [reply lib/util/vector2d.ex:41]: This has docs and tests!

File: lib/util/location.ex
Comment (joshlarson) [reply lib/util/location.ex:130]: I'm thinking maybe... `displacement_from`, and swap the order of the arguments. That way it would be something like:

```elixir
end_loc |> displacement_from(start_loc)
```

And then a similar treatment to `displace`, so that it would become

```elixir
start_loc |> displace_by(vector)
```

(I know that credo won't let us do it _exactly_ like that, but it still might make it easier to tell what's going on.)

Comment (firestack) [reply lib/util/location.ex:130]: I like the idea of reordering the arguments 👍🏻

**Review by firestack — APPROVED** [review]
Maybe @hannahpurcell should have a look over this before merging but I'm good with this 👍🏻 (would be nice to close out the last comment thread though)

File: lib/util/location.ex
Comment (joshlarson) [reply lib/util/location.ex]: [How does this look?](https://github.com/mbta/skate/pull/2510/commits/7bfb9a68b22a5daa540ba86ca751548cf115595a)

Comment (joshlarson) [reply lib/util/location.ex]: [And this to fix some formatting](https://github.com/mbta/skate/pull/2510/commits/4c7295e35b67940a1d354a80b9a6e4bd7cbc4bed)

**Review by hannahpurcell — COMMENTED** [review]
Thank you so much for your patience!! So far, the logic is definitely complex, but making sense with a slow comb-through. I paused before completely finishing because I have a big question that might help things make a little more sense to me.

File: lib/util/location.ex
Comment (joshlarson) [reply lib/util/location.ex:241]: Yes - this is a great point!

File: lib/util/location.ex
Comment (joshlarson) [reply lib/util/location.ex:285]: You're right - I should have said something like `S -> Q is what you get when you project S -> P onto the line containing S -> E`. 

Literally _everything_ except for the lat/long coordinates is vectors, not points (you can kind of tell that in the code, because almost all of the variables inside the function are `%Vector2d{...}`'s, but that's only immediately obvious if you look at the `@spec`'s of the various functions called), but I was fuzzy in my language in the explanation. I shall clean that up before merging.

File: lib/util/location.ex
Comment (joshlarson) [reply lib/util/location.ex:103]: For sure! 

Do you think that "between `to` and `from`". I guess the variable names already make it pretty clear that we're going FROM `from` TO `to` 😂

File: lib/util/location.ex
Comment (joshlarson) [reply lib/util/location.ex:271]: In the paragraph where I refer to `off-segment point`, I'm actually referring to the point itself:
```
Convert both the segment end-point and the off-segment point
into vectors representing their displacements from the segment
start-point.
```

**Comment by joshlarson** [comment]
@hannahpurcell and @firestack I re-worked some of the explanation. Can you give it a once-over and let me know if it's any clearer?

File: lib/util/location.ex
Comment (hannahpurcell) [reply lib/util/location.ex]: Ahhh yes, tracking now. I see my later question about the projection did make this part a little clearer to me

File: lib/util/location.ex
Comment (hannahpurcell) [inline lib/util/location.ex:193]: It seems that the choice of `1000` versus another multiplier might be a bit arbitrary, except that this function needs to be the mathy inverse of the `displacement_from` function. Like, `displacement_from` gets a vector by calculating a scale_factor and multiplying the lat-lng by that factor, and this function undoes that scaling by dividing the lat-lng by the same scale factor. I think this is readable and comprehensible enough, but I can also see these two functions being refactored so that the scale_factor is calculated in a separate function that both `displacement_from` and `displace_by` use. Up to you if you like that or not!

**Review by hannahpurcell — APPROVED** [review]
This is great! Thank you! One nonblocking idea:

File: lib/util/location.ex
Comment (joshlarson) [reply lib/util/location.ex:311]: At first I was all like "ugh, not another letter to remember", but `N` and `Q` are different, and trying to _not_ introduce `N` got even more confusing, I thought.

File: lib/util/location.ex
Comment (joshlarson) [reply lib/util/location.ex:193]: Oh yeah good call. That had bothered me a bit, but then I completely forgot.

Comment (joshlarson) [reply lib/util/location.ex]: I believe that [this](https://github.com/mbta/skate/pull/2510/commits/9d07fd4bcdee050de2ea6720a15d1e42e9e98e94) should address that. I went with `small_increment` to represent `0.001` (and divided by `small_increment` instead of multiplying by `1000`) to make it clearer that it's just a random small number, and also because now you don't have to guess at the relationship between `1000` and `0.001`.

**Pull request merged**
